dutch-auction/src/lib.rs

```diff
@@ -0,0 +1,195 @@
+#![no_std]
+#![feature(const_btree_new)]
+
+use codec::Encode;
+use gstd::{exec::block_timestamp, msg, prelude::*, ActorId};
+use primitive_types::U256;
+
+pub mod state;
+pub use state::{AuctionInfo, State, StateReply};
+
+pub use auction_io::{Action, CreateConfig, Event, InitConfig};
+
+#[derive(Debug, Default)]
+pub struct NFT {
+    pub token_id: U256,
+    pub owner: ActorId,
+    pub contract_id: ActorId,
+}
+
+#[derive(Debug, Default)]
+pub struct Auction {
+    pub owner: ActorId,
+    pub nft: NFT,
+    pub starting_price: U256,
+    pub discount_rate: U256,
+    pub is_active: bool,
+    pub start_at: u64,
+    pub expires_at: u64,
+    pub duration: u64,
+}
+
+static mut AUCTION: Option<Auction> = None;
+
+impl Auction {
+    async fn buy(&mut self) {
+        if !self.is_active {
+            panic!("already bought or auction expired");
+        }
+
+        if block_timestamp() >= self.expires_at {
+            panic!("auction expired");
+        }
+
+        let price = self.token_price().as_u128();
+
+        if msg::value() < price {
+            panic!("value < price");
+        }
+
+        self.is_active = false;
+        let refund = msg::value() - price;
+
+        let _transfer_response: nft_example_io::Event = msg::send_and_wait_for_reply(
+            self.nft.contract_id,
+            nft_example_io::Action::Transfer {
+                to: msg::source(),
+                token_id: self.nft.token_id,
+            },
+            0,
+        )
+        .unwrap()
+        .await
+        .expect("Error in nft transfer");
+
+        msg::send(msg::source(), "", refund).unwrap();
```

Comment:
I'd recommend to use `expect("Explanation")` instead of `unwrap()` when writing a new code.